### PR TITLE
rustc: Prepare to enable ThinLTO by default

### LINKED ISSUE #45320: Tracking issue for enabling multiple CGUs in release mode by default
I'm opening this up to serve as a tracking issue for enabling multiple codegen units in release mode by default. I've written up a [lengthy summary](https://internals.rust-lang.org/t/help-test-out-thinlto/6017) before but the tl;dr; is that multiple codegen units enables us to run optimization/code generation in parallel, making use of all available computing resources often speeding up compilations by more than 2x.

Historically this has not been done due to claims of a loss in performance, but the [recently implemented ThinLTO](https://github.com/rust-lang/rust/pull/44841) is intended to assuage such concerns. The most viable route forward seems to be to enable multiple CGUs and ThinLTO at the same time in release mode.

[Performance summary](https://github.com/rust-lang/rust/issues/45320#issuecomment-337957101)

**Blocking issues:**

* [x] ~~[ThinLTO exposes too many symbols](https://github.com/rust-lang/rust/pull/46374)~~ - [fixed](https://github.com/rust-lang/rust/pull/46382)
* [x] [Enable for rustc on all platforms](https://github.com/rust-lang/rust/pull/45400)
* [x] Enable for libstd on all platforms
  * ~~[first attempt](https://github.com/rust-lang/rust/pull/45529)~~ - ~~[blocked on presumed LLVM bug](https://bugs.llvm.org/show_bug.cgi?id=35140)~~ - ~~[Rust tracking issue](https://github.com/rust-lang/rust/issues/45654)~~ - ~~[fixed](https://github.com/rust-lang/rust/pull/45806)~~
  * ~~[blocked on test failures](https://github.com/rust-lang/rust/pull/45661#issuecomment-341620033)~~ - ~~[current presumed cause of test failures](https://github.com/rust-lang/rust/issues/45511)~~ - ~~[next attempt](https://github.com/rust-lang/rust/pull/46008)~~
* [x] ~~[Possible build-time regressions using multiple CGUs in debug mode](https://github.com/rust-lang/rust/issues/44941)~~ - [couldn't reproduce](https://github.com/rust-lang/rust/issues/44941#issuecomment-338034984)
* [x] ~~[Reported build time regression in rust-doom](https://github.com/rust-lang/rust/issues/45188)~~ - [couldn't reproduce](https://github.com/rust-lang/rust/issues/45188#issuecomment-337369426)
* [x] ~~[ThinLTO broken some MSVC rlibs](https://github.com/rust-lang/rust/issues/45347)~~ - [fixed](https://github.com/rust-lang/rust/pull/45348)
* [x] [More bad debuginfo](https://github.com/rust-lang/rust/issues/46346) - [propoxed fix](https://github.com/rust-lang/rust/pull/46772)

**Potential blockers/bugs:**

* [ ] [ThinLTO broke sanitizers](https://github.com/rust-lang/rust/issues/45220)
* [ ] [ThinLTO for libstd breaks the leak sanitizer](https://github.com/rust-lang/rust/issues/46126)
* [x] [ThinLTO may break debuginfo](https://github.com/rust-lang/rust/issues/45511) - ~~[proposed fix](https://reviews.llvm.org/D39981)~~ - ~~[update to rust](https://github.com/rust-lang/rust/pull/46008)~~
* [ ] [ThinLTO breaks msp430 (and other non-integrated assembler targets)](https://github.com/rust-lang/rust/issues/45836)

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit *almost* enables ThinLTO and multiple codegen units in release mode by
default but is blocked on #46346 now before pulling the trigger.

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @michaelwoerister 

cc @rust-lang/compiler

**Comment by sfackler** [comment]
Is this a blocker? https://github.com/rust-lang/rust/issues/46346

**Comment by alexcrichton** [comment]
Are we sure that's related to ThinLTO? I've tested everything locally and wasn't able to reproduce myself.

**Comment by alexcrichton** [comment]
Urgh it may very well be. I swear to god.

File: src/rustllvm/PassWrapper.cpp [hunk 1]
```diff
@@ -1062,11 +982,27 @@ LLVMRustCreateThinLTOData(LLVMRustThinLTOModule *modules,
     ResolvedODR[ModuleIdentifier][GUID] = NewLinkage;
   };
   thinLTOResolveWeakForLinkerInIndex(Ret->Index, isPrevailing, recordNewLinkage);
+
+  // Here we calculate an `ExportedGUIDs` set for use in the `isExported`
+  // callback below. This callback below will dictate the linkage for all
+  // summaries in the index, and we basically just only want to ensure that dead
+  // symbols are internalized. Otherwise everything that's already external
+  // linkage will stay as external, and internal will stay as internal.
+  std::set<GlobalValue::GUID> ExportedGUIDs;
```
Comment (kennytm) [inline src/rustllvm/PassWrapper.cpp]: Why not use a `std::unordered_set` here.

**Comment by cuviper** [comment]
I see that `LLVMRustThinLTOAvailable` gates on just `LLVM_VERSION_GE(4, 0)`.  How well are you expecting this to work when using an _external_ LLVM-4.0?
(IOW, are there rust-lang/llvm patches that I definitely need here?)

**Comment by alexcrichton** [comment]
@cuviper IIRC https://github.com/rust-lang/llvm/commit/e45c75de1148456a9eb1a67c14a66df4dfb50c94 is the main (and possibly only) ThinLTO related patch (although it fixes a preexisting bug). 

In general though I lose track over time what patch was for what on our fork, although everything that matters is upstream in some form so it's just a matter of time until we update to get it.

**Comment by cuviper** [comment]
Ok, thanks, we already grabbed that one. I'll look back over the branch again later myself. I do appreciate that everything is fixed upstream first!

**Comment by michaelwoerister** [comment]
Awesome! Will review shortly. Am I reading this right, this is still blocked on #46346?

**Comment by alexcrichton** [comment]
@michaelwoerister unfortunately yes :(

File: src/librustc/session/mod.rs [hunk 2]
```diff
@@ -656,30 +656,89 @@ impl Session {
...
+        // always at the same number at all optimization levels.
+        //
+        // As a result 16 was chosen here! Mostly because it was a power of 2
+        // and most benchmarks agreed it was roughly a local optimum. Not very
+        // scientific.
+        16
+    }
+
+    /// Returns whether ThinLTO is enabled for this compilation
+    /// compilation
```
Comment (michaelwoerister) [inline src/librustc/session/mod.rs]: compilation compilation

File: src/libstd/sys_common/backtrace.rs [hunk 3]
```diff
@@ -252,8 +252,26 @@ fn output_fileline(w: &mut Write,
 // Note that this demangler isn't quite as fancy as it could be. We have lots
 // of other information in our symbols like hashes, version, type information,
 // etc. Additionally, this doesn't handle glue symbols at all.
-pub fn demangle(writer: &mut Write, s: &str, format: PrintFormat) -> io::Result<()> {
-    // First validate the symbol. If it doesn't look like anything we're
+pub fn demangle(writer: &mut Write, mut s: &str, format: PrintFormat) -> io::Result<()> {
+    // During ThinLTO LLVM may import and rename internal symbols, so strip out
+    // those endings first as they're on of the last manglings applied to symbol
```
Comment (michaelwoerister) [inline src/libstd/sys_common/backtrace.rs]: on -> one

File: src/rustllvm/PassWrapper.cpp [hunk 4]
```diff
@@ -1004,12 +925,10 @@ LLVMRustCreateThinLTOData(LLVMRustThinLTOModule *modules,
   Ret->Index.collectDefinedGVSummariesPerModule(Ret->ModuleToDefinedGVSummaries);
 
   // Convert the preserved symbols set from string to GUID, this is then needed
-  // for internalization. We use `addPreservedGUID` to include any transitively
-  // used symbol as well.
+  // for internalization.
   for (int i = 0; i < num_symbols; i++) {
-    addPreservedGUID(Ret->Index,
-                     Ret->GUIDPreservedSymbols,
-                     GlobalValue::getGUID(preserved_symbols[i]));
```
Comment (michaelwoerister) [inline src/rustllvm/PassWrapper.cpp]: Was this too conservative before?

**Review by michaelwoerister — COMMENTED** [review]
The PR looks good to me. Let's hope it doesn't bit rot too much until #46346 is resolved.

Can you explain to me why you don't use `addPreservedGUID` anymore?

**Comment by alexcrichton** [comment]
@michaelwoerister ok I've updated the listing of commits and addressed comments. The commits now to *not* enable ThinLTO by default but instead retain retain today's default. The repo is left in a state though to make it a one line change to turn ThinLTO on by default in the future.

To answer your question about `GUIDPreservedSymbols` though there's more info on https://github.com/rust-lang/rust/issues/46374 but it boils down to two things:

* In the first location where `GUIDPreservedSymbols` was used the logic of `addPreservedGUID` wasn't necessary. The same result is computed whether we call `addPreservedGUID` or not.
* In the second location it was *incorrect* to call `addPreservedGUID` in the sense that it added the transitive closure of symbols to that set. That in turn caused https://github.com/rust-lang/rust/issues/46374 where too many symbols were exported (aka all of them) rather than just the set we wanted. By avoiding `addPreservedGUID` and tweaking the definition of `isExported` this bug was fixed though.

**Comment by michaelwoerister** [comment]
Thanks for the explanation!

@bors r+

**Pull request merged**